Core/Src/main.c

```diff
@@ -59,106 +81,133 @@ void SystemClock_Config(void);
 
 /**
-  * @brief  The application entry point.
-  * @retval int
-  */
-int main(void)
-{
-  /* USER CODE BEGIN 1 */
+ * @brief  The application entry point.
+ * @retval int
+ */
+int main(void) {
+    /* USER CODE BEGIN 1 */
+
+    /* USER CODE END 1 */
+
+    /* MCU Configuration--------------------------------------------------------*/
+
+    /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+    HAL_Init();
+
+    /* USER CODE BEGIN Init */
+
+    /* USER CODE END Init */
 
-  /* USER CODE END 1 */
+    /* Configure the system clock */
+    SystemClock_Config();
 
-  /* MCU Configuration--------------------------------------------------------*/
+    /* USER CODE BEGIN SysInit */
 
-  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
-  HAL_Init();
+    /* USER CODE END SysInit */
 
-  /* USER CODE BEGIN Init */
+    /* Initialize all configured peripherals */
+    MX_GPIO_Init();
+    MX_USART2_UART_Init();
+    MX_SPI1_Init();
+    MX_TIM1_Init();
+    /* USER CODE BEGIN 2 */
 
-  /* USER CODE END Init */
+    // Setting CS high to signal no SPI packets
+    HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
 
-  /* Configure the system clock */
-  SystemClock_Config();
+    // Setting duty cycle and starting PWM
+    __HAL_TIM_SET_COMPARE(&htim1, TIM_CHANNEL_1, MIN_PULSE_WIDTH);
+    HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
 
-  /* USER CODE BEGIN SysInit */
+    /* USER CODE END 2 */
 
-  /* USER CODE END SysInit */
+    /* Infinite loop */
+    /* USER CODE BEGIN WHILE */
+    while (1) {
+        // Setting CS low to signal start of SPI packet
+        HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
 
-  /* Initialize all configured peripherals */
-  MX_GPIO_Init();
-  MX_USART2_UART_Init();
-  /* USER CODE BEGIN 2 */
+        // I should probably check the return value here
+        HAL_SPI_TransmitReceive(&hspi1, (uint8_t*) &ADC_TX_BYTES,
+                (uint8_t*) &adc_out, 2, SPI_COMM_TIMEOUT);
 
-  /* USER CODE END 2 */
+        // Setting CS high to signal end of SPI packet
+        HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
 
-  /* Infinite loop */
-  /* USER CODE BEGIN WHILE */
-  while (1)
-  {
-    /* USER CODE END WHILE */
+        // Masking ADC output to avoid taking any values not part of measurement
+        adc_out &= ADC_RX_MASK;
 
-    /* USER CODE BEGIN 3 */
-  }
-  /* USER CODE END 3 */
+        // Linear interpolation
+        // There's probably some assembly instruction that will multiply two 16 uint16_t to a uint32_t
+        // If this was x86 there'd be an instruction that multiplies two pairs of 16 bit numbers into a 32 bit number,
+        // adds the products, then shifts back to a 16 bit
+        uint16_t pulse_width =
+                (uint16_t) ((uint32_t) (ADC_MAX_READING - adc_out)
+                        * MIN_PULSE_WIDTH + (uint32_t) adc_out * MAX_PULSE_WIDTH)
+                        >> ADC_READING_LEN;
```

Comment:
The logic here is a bit confusing. I think everything inside your big bracket could be simplfied to adc_out * min_pulse width + adc_max * min_pulse_width. And the bit shifting here is equivalent to dividing the number by 2^10? I think though your approach is different from other people's approach, it still derive to the correct solution eventually